### PR TITLE
Deprecate find_plugin in boundary_composition/interface.h

### PR DESCRIPTION
**Pull request by joeschools**
Getting started on #2260. Looking for tests and reviews on this one before the next five on the list.
Deprecated find_boundary_composition_model in favor of has_matching_boundary_composition_model and get_matching_boundary_composition_model, and added tests following the examples of #2148, #2255, and #2261.

### CODE REVIEW

File: include/aspect/boundary_composition/interface.h [hunk 1]
```diff
@@ -333,4 +376,4 @@ namespace aspect
 }
 
 
-#endif
+#endif
```
Comment (tjhei) [inline include/aspect/boundary_composition/interface.h]: add a newline

File: include/aspect/boundary_composition/interface.h [hunk 2]
```diff
@@ -27,9 +27,13 @@
 #include <aspect/utilities.h>
 #include <aspect/simulator_access.h>
 
+#include <deal.II/base/std_cxx11/shared_ptr.h>
```
Comment (tjhei) [inline include/aspect/boundary_composition/interface.h]: we can now use ``std::shared_ptr``, see #2378. Can you please remove this header and replace the type below?

File: include/aspect/boundary_composition/interface.h [hunk 3]
```diff
@@ -215,15 +219,26 @@ namespace aspect
...
+         * Go through the list of all boundary composition models that have been selected
+         * in the input file (and are consequently currently active) and see
+         * if one of them has the type specified by the template
+         * argument or can be casted to that type. If so, return a reference
+         * to it. If no boundary composition model is active that matches the given type,
+         * throw an exception.
          */
         template <typename BoundaryCompositionType>
-        BoundaryCompositionType *
-        find_boundary_composition_model () const;
```
Comment (tjhei) [inline include/aspect/boundary_composition/interface.h]: You are removing this function. Has this not been used/tested before?

Comment (joeschools) [reply include/aspect/boundary_composition/interface.h]: It was seemingly never used before, but I added it back in with a deprecation warning

**Comment by gassmoeller** [comment]
You will need to update the test result, let me know if you need help with that. Apart from that this is ready.

**Comment by gassmoeller** [comment]
Looks good. Thanks a lot for the cleanup! :+1:

**Pull request merged**